py/server/deephaven/table_factory.py

```diff
@@ -335,3 +340,88 @@ def ring_table(parent: Table, capacity: int, initialize: bool = True) -> Table:
         return Table(j_table=_JRingTableTools.of(parent.j_table, capacity, initialize))
     except Exception as e:
         raise DHError(e, "failed to create a ring table.") from e
+
+
+def function_generated_table(table_generator: Callable[[], Table],
+           source_tables: Union[Table, List[Table]] = None,
+           refresh_interval_ms: int = None,
+           exec_ctx: ExecutionContext = None) -> Table:
+    """
+    Creates an abstract table that is generated by running the table_generator() function. The function will first be
+    run to generate the table when this method is called, then subsequently either (a) whenever one of the
+    'source_tables' ticks or (b) after refresh_interval_ms have elapsed. Either 'refresh_interval_ms' or
+    'source_tables' must be set (but not both).
+
+    Function-generated tables can be used to produce dynamic tables from sources outside Deephaven. For example,
+    function-generated tables can create tables that are produced by arbitrary Python logic (including using Pandas or
+    numpy). They can also be used to retrieve data from external sources (such as files or websites).
+
+    The table definition must not change between invocations of the 'table_generator' function.
+
+    Note that any tables used by the 'table_generator' function *MUST* be specified in the 'source_tables'.
+
+    Args:
+        table_generator (Callable[[], Table]): The table generator function. This function must return a Table.
+        source_tables (Union[Table, List[Table]]): Source tables used by the 'table_generator' function. The
+            'table_generator' is rerun when any of these tables tick.
+        refresh_interval_ms (int): Interval (in milliseconds) at which the 'table_generator' function is rerun.
+        exec_ctx (ExecutionContext) -> Table): Optionally, a custom execution context. If 'None', the current
+            execution context is used. If there is no current execution context, a ValueError is raised.
+
+    Returns:
+        A Table that is automatically regenerated when any of the source_tables tick. This table will be regenerated
+        *after* any changes to the source_tables have been processed.
```

Comment:
```suggestion
    Returns:
        a new table
        
    Raises:
        DHError
```
This is more consistent with the existing docs.  The first sentence is covered already.  If we need second sentence, it should be moved to the body.